server/lib/api/parties.ts

```diff
@@ -0,0 +1,182 @@
+import { USERNAME_MAXLENGTH, USERNAME_MINLENGTH, USERNAME_PATTERN } from '../../../common/constants'
+import { PARTIES } from '../../../common/flags'
+import { featureEnabled } from '../flags/feature-enabled'
+import users from '../models/users'
+import PartyService, {
+  PartyServiceError,
+  PartyServiceErrorCode,
+  PartyUser,
+} from '../parties/party-service'
+import ensureLoggedIn from '../session/ensure-logged-in'
+import createThrottle from '../throttle/create-throttle'
+import throttleMiddleware from '../throttle/middleware'
+import { joiValidator } from '../validation/joi-validator'
+
+const partyService = container.resolve(PartyService)
+
+const invitesThrottle = createThrottle('partyInvites', {
+  rate: 40,
+  burst: 60,
+  window: 60000,
+})
+
+const partyThrottle = createThrottle('parties', {
+  rate: 20,
+  burst: 40,
+  window: 60000,
+})
+
+const clientIdSchema = Joi.object().keys({
+  clientId: Joi.string().required(),
+})
+
+const invitesPostSchema = clientIdSchema.keys({
+  targets: Joi.array()
+    .items(
+      Joi.string()
+        .min(USERNAME_MINLENGTH)
+        .max(USERNAME_MAXLENGTH)
+        .pattern(USERNAME_PATTERN)
+        .required(),
+    )
+    .min(1)
+    .required(),
+})
+
+const partyIdSchema = Joi.object().keys({
+  partyId: Joi.string().required(),
+})
+
+const declineParamsSchema = partyIdSchema.keys({
+  targetId: Joi.number(),
+})
+
+export default function (router: Router) {
+  router
+    .use(ensureLoggedIn)
+    .post(
+      '/invites',
+      featureEnabled(PARTIES),
+      throttleMiddleware(invitesThrottle, ctx => ctx.session!.userId),
+      joiValidator({ body: invitesPostSchema }),
+      invite,
+    )
+    .delete(
+      '/invites/:partyId/:targetId',
+      featureEnabled(PARTIES),
+      throttleMiddleware(invitesThrottle, ctx => ctx.session!.userId),
+      joiValidator({ params: declineParamsSchema }),
+      decline,
+    )
+    .post(
+      '/:partyId',
+      featureEnabled(PARTIES),
+      throttleMiddleware(partyThrottle, ctx => ctx.session!.userId),
+      joiValidator({ params: partyIdSchema }),
+      accept,
+    )
+}
+
+function isPartyServiceError(error: Error): error is PartyServiceError {
+  return error.hasOwnProperty('code')
+}
+
+function convertPartyServiceError(err: Error) {
+  if (!isPartyServiceError(err)) {
+    throw err
+  }
+
+  switch (err.code) {
+    case PartyServiceErrorCode.PartyNotFound:
+      throw new httpErrors.NotFound(err.message)
+    case PartyServiceErrorCode.InsufficientPermissions:
+      throw new httpErrors.Forbidden(err.message)
+    case PartyServiceErrorCode.PartyFull:
+      throw new httpErrors.Conflict(err.message)
+    case PartyServiceErrorCode.UserOffline:
+      throw new httpErrors.Unauthorized(err.message)
+    case PartyServiceErrorCode.InvalidAction:
+      throw new httpErrors.BadRequest(err.message)
+    default:
+      assertUnreachable(err.code as never)
```

Comment:
Don't do this cast. At best this is casting something that already has the type `never`, at worst it makes this assertion not work.

`assertUnreachable` works because TypeScript is smart enough to look at the cases of the switch and realize that when all possible values are exhausted by case statements, the type of the switch'd value is non-existent (e.g. `never`). If you call `assertUnreachable` with a value when that value *hasn't* been auto-narrowed to `never`, the compilation will fail. But that is *the entire point*: it's meant to tell you if someone adds a value to the enum without also updating this switch.

So, never cast the value you pass into it